> 🤖 **Posted autonomously by Claude Code acting on the user's behalf.** Triaging this rebased PR's CI; classifying the NPU failure below and re-running it. Please push back if any conclusion is off.

**Non-CUDA (NPU/Ascend) failure — classified as a perf-threshold flake, unrelated to this PR.**

- Failed job: `stage-b-test-1-npu-a2 (0)` → cascaded into `pr-test-npu-finish`. Sibling shards (`(1)`, `stage-b-test-2/4/16-npu`, `multimodal-gen-npu`) all passed.
- Test: `test/registered/ascend/basic_function/quant/test_npu_w8a8_quantization.py`
- Fingerprint:
  ```
  AssertionError: Output throughput of .../Qwen2.5-0.5B-Instruct-quantized.w8a8 is 635.55, is lower than 700
  (retry: 542.18, also < 700)
  ```
- Why it's not this PR: this is a **throughput threshold** assertion (≥700 tok/s), not a correctness check. The PR is a value-equivalent scheduler refactor (merge `extend_input_len`/`fill_len` → `Req.extend_range`; `extend_logprob_start_len` → a pure function; fill-id reads route through the existing `full_untruncated_fill_ids` field, same as `main`). It touches no Ascend code, and reads are if anything cheaper than the prior getter-based form. The ~9–23% shortfall with large run-to-run variance (635→542) is NPU-runner noise.
- Next: re-running the NPU test to clear the gate. The CUDA lanes (the hard gate) are green so far.

--- a/python/sglang/srt/disaggregation/decode_schedule_batch_mixin.py
+++ b/python/sglang/srt/disaggregation/decode_schedule_batch_mixin.py
@@ -74,8 +74,6 @@ class ScheduleBatchDisaggregationDecodeMixin:
             req.is_retracted = False
             pre_lens.append(pre_len)
 
-        extend_input_logprob_token_ids = None
-
         # Set fields
         self.input_ids = torch.tensor(
             sum(input_ids, array("q")), dtype=torch.int32, device=self.device
@@ -99,9 +97,8 @@ class ScheduleBatchDisaggregationDecodeMixin:
         self.extend_num_tokens = extend_num_tokens
         self.prefix_lens = [len(r.prefix_indices) for r in reqs]
         self.extend_lens = [r.extend_range.length for r in reqs]
-        # Prebuilt reqs already had input logprobs computed on the prefill server.
-        self.extend_logprob_start_lens = [0] * len(reqs)
-        self.extend_input_logprob_token_ids = extend_input_logprob_token_ids
+        self.extend_logprob_start_lens = None
+        self.extend_input_logprob_token_ids = None
         self.multimodal_inputs = [r.multimodal_inputs for r in reqs]
 
         # Build sampling info

--- a/python/sglang/srt/disaggregation/decode_schedule_batch_mixin.py
+++ b/python/sglang/srt/disaggregation/decode_schedule_batch_mixin.py
@@ -73,7 +73,6 @@ class ScheduleBatchDisaggregationDecodeMixin:
                 req.already_computed = seq_len
             req.is_retracted = False
             pre_lens.append(pre_len)
-            req.extend_logprob_start_len = 0
 
         extend_input_logprob_token_ids = None
 
@@ -100,7 +99,8 @@ class ScheduleBatchDisaggregationDecodeMixin:
         self.extend_num_tokens = extend_num_tokens
         self.prefix_lens = [len(r.prefix_indices) for r in reqs]
         self.extend_lens = [r.extend_range.length for r in reqs]
-        self.extend_logprob_start_lens = [r.extend_logprob_start_len for r in reqs]
+        # Prebuilt reqs already had input logprobs computed on the prefill server.
+        self.extend_logprob_start_lens = [0] * len(reqs)
         self.extend_input_logprob_token_ids = extend_input_logprob_token_ids
         self.multimodal_inputs = [r.multimodal_inputs for r in reqs]
 

--- a/python/sglang/srt/managers/schedule_batch.py
+++ b/python/sglang/srt/managers/schedule_batch.py
@@ -1637,21 +1637,10 @@ def compute_extend_logprob_start_len(
     extend_len: int,
     full_untruncated_fill_len: int,
 ) -> int:
-    """Relative position within an extend window where input-logprob computation starts.
-
-    Pure function of the extend forward inputs; it holds no Req state. Callers snapshot
-    the result at forward-assembly time into ScheduleBatch.extend_logprob_start_lens,
-    because the value is bound to one extend forward, not to live Req state (under
-    overlap scheduling the Req's prefix_indices/extend_range/logprob_start_len are
-    overwritten by the next round before the current forward's output is processed).
-
-    Key variables:
-    - logprob_start_len: absolute position in the full sequence where logprob starts
-      (-1 means "no input logprobs", i.e. start at the very end of the sequence).
-    - prefix_len: number of cached prefix tokens (extend window starts here).
-    - extend_len: number of tokens processed in this extend window.
-    - full_untruncated_fill_len: full sequence length, used to resolve the -1 sentinel.
-    """
+    # Key variables:
+    # - logprob_start_len: Absolute position in full sequence where logprob computation begins
+    # - extend_logprob_start_len: Relative position within current extend batch where logprob computation begins
+    # - extend_input_len: Number of tokens that need to be processed in this extend batch
     if logprob_start_len == -1:
         resolved_start = full_untruncated_fill_len
     else:
@@ -2030,9 +2019,6 @@ class ScheduleBatch(ScheduleBatchDisaggregationDecodeMixin):
         orig_seq_lens = [max(r.extend_range.end, len(r.origin_input_ids)) for r in reqs]
         prefix_lens = [len(r.prefix_indices) for r in reqs]
         extend_lens = [r.extend_range.length for r in reqs]
-        # Snapshot the per-req input-logprob start (relative to the extend window).
-        # Encoder-decoder stripping adjusts this list in place in
-        # prepare_encoder_info_extend; do not recompute it afterwards.
         extend_logprob_start_lens = [
             compute_extend_logprob_start_len(
                 logprob_start_len=r.logprob_start_len,
@@ -2190,12 +2176,12 @@ class ScheduleBatch(ScheduleBatchDisaggregationDecodeMixin):
                 ]
                 extend_input_logprob_token_ids.extend(logprob_token_ids)
 
-                # We will need extend_len - extend_logprob_start_len number of
+                # We will need req.extend_range.length - extend_logprob_start_lens[i] number of
                 # tokens, and logprob_token_ids is for input logprob, so pad the rest of them by 0.
                 extend_input_logprob_token_ids.extend(
                     [0]
                     * (
-                        extend_lens[i]
+                        req.extend_range.length
                         - extend_logprob_start_lens[i]
                         - len(logprob_token_ids)
                     )

--- a/python/sglang/srt/managers/schedule_batch.py
+++ b/python/sglang/srt/managers/schedule_batch.py
@@ -844,8 +844,6 @@ class Req(ReqDllmMixin):
         # Prefix info
         # The indices to kv cache for the shared prefix.
         self.prefix_indices: torch.Tensor = torch.empty((0,), dtype=torch.int64)
-        # The relative logprob_start_len in an extend batch
-        self.extend_logprob_start_len = 0
         # TODO(ispobock): rename to last_device_node
         self.last_node: Any = None
         self.last_host_node: Any = None
@@ -1098,7 +1096,6 @@ class Req(ReqDllmMixin):
 
     def set_extend_range(self, start: int, end: int) -> None:
         self.extend_range = Range(start, end)
-        self._recompute_extend_logprob_start_len()
 
     def get_fill_ids(self) -> array:
         return self.full_untruncated_fill_ids[: self.extend_range.end]
@@ -1458,7 +1455,6 @@ class Req(ReqDllmMixin):
         self.input_token_logprobs = None
         self.temp_input_top_logprobs_val = None
         self.temp_input_top_logprobs_idx = None
-        self.extend_logprob_start_len = 0
         self.inflight_middle_chunks = 0
         self.mamba_pool_idx = None
         self.mamba_ping_pong_track_buffer = None
@@ -1523,23 +1519,6 @@ class Req(ReqDllmMixin):
         )
         logger.info(f"{prefix}: {self.time_stats.convert_to_duration()}")
         self.has_log_time_stats = True
-
-    def _recompute_extend_logprob_start_len(self):
-        # Setting extend_input_len and computing the relative logprob_start_len in an extend batch
-        #
-        # Key variables:
-        # - logprob_start_len: Absolute position in full sequence where logprob computation begins
-        # - extend_logprob_start_len: Relative position within current extend batch where logprob computation begins
-        # - extend_input_len: Number of tokens that need to be processed in this extend batch
-        if self.logprob_start_len == -1:
-            logprob_start_len = len(self.full_untruncated_fill_ids)
-        else:
-            # logprob_start_len should be at least the length of the prefix indices
-            logprob_start_len = max(self.logprob_start_len, len(self.prefix_indices))
-        self.extend_logprob_start_len = min(
-            logprob_start_len - len(self.prefix_indices),
-            self.extend_range.length,
-        )
 
     def set_finish_with_abort(self, error_msg: str):
         if get_tensor_model_parallel_rank() == 0:
@@ -1649,6 +1628,36 @@ def retract_all(
             hisparse_coordinator=hisparse_coordinator,
         )
     return retracted_reqs
+
+
+def compute_extend_logprob_start_len(
+    *,
+    logprob_start_len: int,
+    prefix_len: int,
+    extend_len: int,
+    full_untruncated_fill_len: int,
+) -> int:
+    """Relative position within an extend window where input-logprob computation starts.
+
+    Pure function of the extend forward inputs; it holds no Req state. Callers snapshot
+    the result at forward-assembly time into ScheduleBatch.extend_logprob_start_lens,
+    because the value is bound to one extend forward, not to live Req state (under
+    overlap scheduling the Req's prefix_indices/extend_range/logprob_start_len are
+    overwritten by the next round before the current forward's output is processed).
+
+    Key variables:
+    - logprob_start_len: absolute position in the full sequence where logprob starts
+      (-1 means "no input logprobs", i.e. start at the very end of the sequence).
+    - prefix_len: number of cached prefix tokens (extend window starts here).
+    - extend_len: number of tokens processed in this extend window.
+    - full_untruncated_fill_len: full sequence length, used to resolve the -1 sentinel.
+    """
+    if logprob_start_len == -1:
+        resolved_start = full_untruncated_fill_len
+    else:
+        # logprob_start_len should be at least the length of the prefix indices
+        resolved_start = max(logprob_start_len, prefix_len)
+    return min(resolved_start - prefix_len, extend_len)
 
 
 def _compute_chunked_req_next_prompt_token(
@@ -2004,9 +2013,6 @@ class ScheduleBatch(ScheduleBatchDisaggregationDecodeMixin):
                 req.extend_range = req.extend_range._replace(
                     start=req.extend_range.start + encoder_len
                 )
-                req.extend_logprob_start_len = max(
-                    0, req.extend_logprob_start_len - encoder_len
-                )
             req.logprob_start_len = max(req.logprob_start_len, encoder_len)
 
     def prepare_for_extend(self):
@@ -2024,6 +2030,18 @@ class ScheduleBatch(ScheduleBatchDisaggregationDecodeMixin):
         orig_seq_lens = [max(r.extend_range.end, len(r.origin_input_ids)) for r in reqs]
         prefix_lens = [len(r.prefix_indices) for r in reqs]
         extend_lens = [r.extend_range.length for r in reqs]
+        # Snapshot the per-req input-logprob start (relative to the extend window).
+        # Encoder-decoder stripping adjusts this list in place in
+        # prepare_encoder_info_extend; do not recompute it afterwards.
+        extend_logprob_start_lens = [
+            compute_extend_logprob_start_len(
+                logprob_start_len=r.logprob_start_len,
+                prefix_len=prefix_lens[i],
+                extend_len=extend_lens[i],
+                full_untruncated_fill_len=len(r.full_untruncated_fill_ids),
+            )
+            for i, r in enumerate(reqs)
+        ]
 
         _pin = is_pin_memory_available(self.device)
         # Stay on pinned CPU; H2D is deferred to forward stream via
@@ -2172,13 +2190,13 @@ class ScheduleBatch(ScheduleBatchDisaggregationDecodeMixin):
                 ]
                 extend_input_logprob_token_ids.extend(logprob_token_ids)
 
-                # We will need req.extend_range.length - req.extend_logprob_start_len number of
+                # We will need extend_len - extend_logprob_start_len number of
                 # tokens, and logprob_token_ids is for input logprob, so pad the rest of them by 0.
                 extend_input_logprob_token_ids.extend(
                     [0]
                     * (
-                        req.extend_range.length
-                        - req.extend_logprob_start_len
+                        extend_lens[i]
+                        - extend_logprob_start_lens[i]
                         - len(logprob_token_ids)
                     )
                 )
@@ -2237,7 +2255,7 @@ class ScheduleBatch(ScheduleBatchDisaggregationDecodeMixin):
             self.top_logprobs_nums = [r.logprob.top_logprobs_num for r in reqs]
             self.token_ids_logprobs = [r.logprob.token_ids_logprob for r in reqs]
 
-        self.extend_logprob_start_lens = [r.extend_logprob_start_len for r in reqs]
+        self.extend_logprob_start_lens = extend_logprob_start_lens
         self.extend_input_logprob_token_ids = extend_input_logprob_token_ids
 
         if get_global_server_args().enable_mamba_extra_buffer():

--- a/python/sglang/srt/managers/scheduler.py
+++ b/python/sglang/srt/managers/scheduler.py
@@ -3340,9 +3340,18 @@ class Scheduler(
                     req.extend_range.length if req.extend_range is not None else 0
                     for req in batch.reqs
                 ]
-                batch_result.extend_logprob_start_len_per_req = [
-                    req.extend_logprob_start_len for req in batch.reqs
-                ]
+                # extend_logprob_start_lens is the forward-time snapshot (already
+                # encoder-adjusted). It is aligned with batch.reqs only right after
+                # prepare_for_extend/mix_with_running; filter_batch/merge_batch do not
+                # maintain it, so a decode batch may carry a stale, misaligned list.
+                # That is harmless (decode never consumes it) but we fall back to zeros
+                # to keep the per-req list aligned with batch.reqs.
+                start_lens = batch.extend_logprob_start_lens
+                batch_result.extend_logprob_start_len_per_req = (
+                    list(start_lens)
+                    if start_lens is not None and len(start_lens) == len(batch.reqs)
+                    else [0] * len(batch.reqs)
+                )
             else:
                 batch_result.extend_input_len_per_req = None
                 batch_result.extend_logprob_start_len_per_req = None

--- a/python/sglang/srt/managers/scheduler.py
+++ b/python/sglang/srt/managers/scheduler.py
@@ -3340,17 +3340,8 @@ class Scheduler(
                     req.extend_range.length if req.extend_range is not None else 0
                     for req in batch.reqs
                 ]
-                # extend_logprob_start_lens is the forward-time snapshot (already
-                # encoder-adjusted). It is aligned with batch.reqs only right after
-                # prepare_for_extend/mix_with_running; filter_batch/merge_batch do not
-                # maintain it, so a decode batch may carry a stale, misaligned list.
-                # That is harmless (decode never consumes it) but we fall back to zeros
-                # to keep the per-req list aligned with batch.reqs.
-                start_lens = batch.extend_logprob_start_lens
                 batch_result.extend_logprob_start_len_per_req = (
-                    list(start_lens)
-                    if start_lens is not None and len(start_lens) == len(batch.reqs)
-                    else [0] * len(batch.reqs)
+                    batch.extend_logprob_start_lens
                 )
             else:
                 batch_result.extend_input_len_per_req = None

--- a/test/registered/unit/managers/test_prefill_adder.py
+++ b/test/registered/unit/managers/test_prefill_adder.py
@@ -79,7 +79,6 @@ class TestPrefillAdder(CustomTestCase):
         req.priority = priority
         req.prefix_indices = []
         req.full_untruncated_fill_ids = []
-        req.extend_logprob_start_len = 0
         req.output_ids = [0] * output_len
         req.sampling_params = SimpleNamespace(max_new_tokens=max_new_tokens)
         req.time_stats = SimpleNamespace(wait_queue_entry_time=wait_time)

